File: Java (client)/ClientServerWithJava/Server/src/main/java/ro/mpp2024/ShowServicesImpl.java
package ro.mpp2024;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ShowServicesImpl implements IService{
    private IRepository<Integer, Match> matchRepository;
    private IRepository<Integer, Team> teamRepository;
    private IRepository<Integer, Ticket> ticketRepository;
    private IRepository<Integer, User> userRepository;
    private Map<String, Observer> loggedClients;
    private final int defaultThreadsNo=5;

    public ShowServicesImpl(IRepository<Integer, Match> matchRepository, IRepository<Integer, Team> teamRepository, IRepository<Integer, Ticket> ticketRepository, IRepository<Integer, User> userRepository) {
        this.matchRepository = matchRepository;
        this.teamRepository = teamRepository;
        this.ticketRepository = ticketRepository;
        this.userRepository = userRepository;
        loggedClients= new ConcurrentHashMap<>();
    }
    @Override
    public synchronized Optional<User> login(String username, String password, Observer client) {
        var utilizator = getAccountByName(username);
        if (utilizator!= null) {
            var parola = utilizator.getPassword();
            if(!password.equals(parola)){
                throw new IllegalArgumentException("Parola incorecta");
            }
            else {
                if (loggedClients.get(username) != null)
                    throw new IllegalArgumentException("Utilizatorul este deja autentificat");
                loggedClients.put(username, client);
                return Optional.of(utilizator);
            }
        }
        return Optional.empty();
    }

    public synchronized User getAccountByName(String username){
        Iterable<User> allusers=userRepository.findAll();
        for (User us:allusers){
            if (Objects.equals(us.getName(), username))
            {
                return us;
            }
        }
        return null;
    }

    @Override
    public synchronized void logout(String username){
        if (loggedClients.get(username) == null)
            throw new IllegalArgumentException("Utilizatorul nu este autentificat");
        loggedClients.remove(username);
    }

    @Override
    public synchronized MatchDTO getMatch(int id){
        Match match = matchRepository.findOne(id);
        MatchDTO matchDTO = new MatchDTO(match.getId(), match.getType(), match.getTeam1ID(), match.getTeam2ID(), match.getAvailableSeats(), match.getSeatPrice());
        return matchDTO;
    }

    @Override
    public synchronized Collection<String> getListMatches(String search) {
        Collection<String> collection = new ArrayList<>();
        Iterable<Match> matches = matchRepository.findAll();
        matches.forEach(match -> {
            String toCollection;
            if(match.getAvailableSeats() == 0)
            {
                toCollection = match.getId() + " | " + match.getType() + ": " + teamRepository.findOne(match.getTeam1ID()).getName() + " VS " + teamRepository.findOne(match.getTeam2ID()).getName() + " | Seat Price: " + match.getSeatPrice() + " | SOLD OUT";}
            else{
                toCollection = match.getId() + " | " + match.getType() + ": " + teamRepository.findOne(match.getTeam1ID()).getName() + " VS " + teamRepository.findOne(match.getTeam2ID()).getName() + " | Seat Price: " + match.getSeatPrice() + " | Available Seats: " + match.getAvailableSeats();}
            if(search == null || toCollection.contains(search))
                collection.add(toCollection);
        });
        return collection;
    }

    @Override
    public synchronized boolean addTicket(BuyTicketsDTO tick){
        Ticket ticket = new Ticket(tick.getMatchID(), tick.getPrice(), tick.getClientName(), tick.getNumberOfSeats());
        Ticket bool = ticketRepository.save(ticket);
        Match match = matchRepository.findOne(tick.getMatchID());
        match.setAvailableSeats(match.getAvailableSeats()-tick.getNumberOfSeats());
        matchRepository.update(match);
        notifyClients();
        return bool != null;
    }

    private void notifyClients(){
        ExecutorService executor = Executors.newFixedThreadPool(defaultThreadsNo);
        for(var client : loggedClients.values()){
            if(client == null)
                continue;
            executor.execute(client::update);
        }
    }
}

File: C# (server)/ClientServerWithJava/Server/ShowServicesImpl.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Service;
using Model;
using Persistence;

namespace Server
{
    public class ShowServicesImpl : IService
    {
        private IRepository<int, Match> matchRepository;
        private IRepository<int, Team> teamRepository;
        private IRepository<int, Ticket> ticketRepository;
        private IRepository<int, User> userRepository;
        private Dictionary<string, Observer> loggedClients;
        private readonly int defaultThreadsNo = 5;
        private readonly object lockObject = new object();

        public ShowServicesImpl(IRepository<int, Match> matchRepository, IRepository<int, Team> teamRepository, IRepository<int, Ticket> ticketRepository, IRepository<int, User> userRepository)
        {
            this.matchRepository = matchRepository;
            this.teamRepository = teamRepository;
            this.ticketRepository = ticketRepository;
            this.userRepository = userRepository;
            loggedClients = new Dictionary<string, Observer>();
        }

        public User Login(string username, string password, Observer client)
        {
                var utilizator = GetAccountByName(username);
                if (utilizator != null)
                {
                    var parola = utilizator.password;
                    if (password != parola)
                    {
                        throw new ArgumentException("Parola incorecta");
                    }
                    else
                    {
                        if (loggedClients.ContainsKey(username))
                            throw new ArgumentException("Utilizatorul este deja autentificat");
                        loggedClients[username] = client;
                        return utilizator;
                    }
                }

                return null;
        }

        public User GetAccountByName(string username)
        {
            lock (lockObject)
            {
                var allUsers = userRepository.findAll();
                foreach (var us in allUsers)
                {
                    if (us.name == username)
                    {
                        return us;
                    }
                }

                return null;
            }
        }

        public void Logout(string username)
        {
            lock (lockObject)
            {
                if (!loggedClients.ContainsKey(username))
                    throw new ArgumentException("Utilizatorul nu este autentificat");
                loggedClients.Remove(username);
            }
        }

        public MatchDTO GetMatch(int id)
        {
            lock (lockObject)
            {
                var match = matchRepository.findOne(id);
                return new MatchDTO(match.id, match.type, match.team1ID, match.team2ID, match.availableSeats,
                    match.seatPrice);
            }
        }

        public ICollection<string> GetListMatches(string search)
        {
            lock (lockObject)
            {
                var collection = new List<string>();
                var matches = matchRepository.findAll();
                foreach (var match in matches)
                {
                    string toCollection;
                    if (match.availableSeats == 0)
                    {
                        toCollection =
                            $"{match.id} | {match.type}: {teamRepository.findOne(match.team1ID).name} VS {teamRepository.findOne(match.team2ID).name} | Seat Price: {match.seatPrice} | SOLD OUT";
                    }
                    else
                    {
                        toCollection =
                            $"{match.id} | {match.type}: {teamRepository.findOne(match.team1ID).name} VS {teamRepository.findOne(match.team2ID).name} | Seat Price: {match.seatPrice} | Available Seats: {match.availableSeats}";
                    }

                    if (search == null || toCollection.Contains(search))
                        collection.Add(toCollection);
                }

                return collection;
            }
        }

        public bool AddTicket(BuyTicketsDTO tick)
        {
            lock (lockObject)
            {
                var ticket = new Ticket(tick.MatchID, tick.Price, tick.ClientName, tick.NumberOfSeats);
                var boolTicket = ticketRepository.save(ticket);
                var match = matchRepository.findOne(tick.MatchID);
                match.availableSeats -= tick.NumberOfSeats;
                matchRepository.update(match);
                NotifyClients();
                return boolTicket != null;
            }
        }

        private void NotifyClients()
        {
            lock (lockObject)
            {
                var tasks = new List<Task>();
                foreach (var client in loggedClients.Values)
                {
                    if (client == null)
                        continue;
                    tasks.Add(Task.Run(() => client.Update()));
                }

                Task.WaitAll(tasks.ToArray());
            }
        }
    }
}
